### PR TITLE
Event based incapacitated, able_to_run

### PR DESCRIPTION
**Pull request by LemonInTheDark**
## About The Pull Request

Pulled out from #79498 

[Sets up event based incapacitated](https://github.com/tgstation/tgstation/pull/82635/commits/b86d5f0d73b2060ee3d851bf07d6e02ef799cc97)
[b86d5f0](https://github.com/tgstation/tgstation/pull/82635/commits/b86d5f0d73b2060ee3d851bf07d6e02ef799cc97)

Rather then calling a proc every time we want to know if we're incap or not, let's store all possible responses as a bitflag and update a var whenever we change.

This way we can react to incap changes in an event based way!

[Implements event based able_to_run(), which should save a good portion off ai work](https://github.com/tgstation/tgstation/pull/82635/commits/2fb26b4a648fd142f3594f63d1f466b220cb65e5)
[2fb26b4](https://github.com/tgstation/tgstation/pull/82635/commits/2fb26b4a648fd142f3594f63d1f466b220cb65e5)

[Refactors incapacitated to support multiple flags correctly](https://github.com/tgstation/tgstation/pull/82635/commits/961c320282e1c1b7103fe7fdf8f6f20564df88c4)
[961c320](https://github.com/tgstation/tgstation/pull/82635/commits/961c320282e1c1b7103fe7fdf8f6f20564df88c4)

Includes a macro to make it easier to reason about.
Also moved all my signal stuff to init_signals, rip setup_incapacitated()

## Why It's Good For The Game

Basic mobs are slow, let's do some work to make their "do this constantly" stuff less horrible.

### What is Lost
This reworks incapacitated handling slightly.
We go from
```
mob.incapacitated()
mob.incipacitated(IGNORE_GRAB)
```
to
```
mob.incapacitated
INCAPACITATED_IGNORING(mob, INCAPACITATED_GRAB)
```

Unsure if I am happy with this api, post opinions in comments please

Related:
Because incapacitated and able_to_run are event based, this virally forces anything that feeds into it to be event based too.
I think this is overall positive but it does place a higher cost on adding to it, and a higher knowledge gate.

### CODE REVIEW

**Comment by Iamgoofball** [comment]
lets fuckin GOOOOOOOOOO

File: code/modules/mob/living/basic/bots/_bots.dm [hunk 1]
```diff
@@ -143,6 +143,11 @@ GLOBAL_LIST_INIT(command_strings, list(
 	ai_controller.set_blackboard_key(BB_RADIO_CHANNEL, radio_channel)
 	update_appearance()
 
+/mob/living/basic/bot/proc/set_mode_flags(mode_flags)
+	SHOULD_CALL_PARENT(TRUE)
+	bot_mode_flags = mode_flags
```
[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/modules/mob/living/basic/bots/_bots.dm:148]: maybe. I keep forgetting that option exists tbh

[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/modules/mob/living/basic/bots/_bots.dm:148]: there isn't really a good way of preventing modification outside a proc unfortunately, need a new lint 
PROTECTED basically works but eh

[the remark this answers is not in the document]
Comment (LemonInTheDark) [reply code/modules/mob/living/basic/bots/_bots.dm:148]: yeah this doesn't work cause we want to allow remote reads. Life is pain

**Comment by LemonInTheDark** [comment]
https://github.com/tgstation/tgstation/pull/79498#issuecomment-1867145227 is likely still an issue, will need to do some tming of this to shittest er

**Comment by Iamgoofball** [comment]
stale b gone

**Comment by Iamgoofball** [comment]
stale b gone

**Comment by optimumtact** [comment]
make a new PR when you're ready to finish this potatoe.

**Pull request closed without merging**